### PR TITLE
gl_shader_disk_cache: Improve precompiled shader cache generation speed and size

### PR DESCRIPTION
**Pull request by FreddyFunk**
**This PR is based on #2098 and #2156. Please merge #2098 and #2156 before merging this PR.**

With this PR shaders will not be compressed individually. Rather we store the the shaders uncompressed on compress the whole precompiled shader cache in one piece when its completly generated. This improves the compression ratio.
Also the precompiled gets generated completly in memory with the help of a `virtual binary file`. This eleminates the bottleneck of disk latency and improves speed greatly.

This PR does not change how the precompiled shader cache works on the shader caching level.

The following tests on the precompiled shader cache are done singlethreaded on an i7-8550u and a Samsung NVME SSD 960 EVO. (This is just a small shader cache). Systems with slower HDDs/SSDs will experience a even larger speed boost.

| Compression Algorithm | Compression Level | Size (KB)  | Precompiled shader cache file generation time (seconds)  |
| ------------- |:-------------:| -----:| -----:|
| uncompressed | --- | 79271 |  15.33 |
| LZ4 | --- | 21075 | (currently in master) 15.24 |
| ZStandard | 3 | 12295 |  (currently in this #2098) 14,69 |
| **ZStandard** | 3 | **8809** | (currently in this PR) **4.41** |

### CODE REVIEW

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 1]
```diff
@@ -591,28 +553,34 @@ FileUtil::IOFile ShaderDiskCacheOpenGL::AppendTransferableFile() const {
...
-            return {};
-        }
+    if (file.WriteBytes(compressed.data(), compressed.size()) != compressed.size()) {
+        LOG_ERROR(Render_OpenGL, "Failed to write precompiled cache version in path={}",
+                  precompiled_path);
+        return;
     }
-    return file;
+
+    file.Close();
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp]: Close doesn't need to be called here. It'll happen automatically

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 2]
```diff
@@ -591,28 +553,34 @@ FileUtil::IOFile ShaderDiskCacheOpenGL::AppendTransferableFile() const {
...
+    const auto hash{GetShaderCacheVersionHash()};
+    if (m_precompiled_cache_virtual_file.WriteArray(hash.data(), hash.size()) != hash.size()) {
+        LOG_ERROR(
+            Render_OpenGL,
+            "Failed to write precompiled cache version hash to virtual precompiled cache file");
+    }
+}
+
+void ShaderDiskCacheOpenGL::SaveVirtualPrecompiledFile() {
+
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp]: Unnecessary newline.

File: .gitmodules [hunk 3]
```diff
@@ -37,3 +37,6 @@
 [submodule "discord-rpc"]
     path = externals/discord-rpc
     url = https://github.com/discordapp/discord-rpc.git
+[submodule "externals/zstd"]
+	path = externals/zstd
```
Comment (lioncash) [inline .gitmodules]: We use spaces, not tabs for indentation

File: src/common/virtual_file_util.cpp [hunk 4]
```diff
@@ -0,0 +1,46 @@
+// Copyright 2019 yuzu Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include "virtual_file_util.h"
+
+namespace Common::VirtualFileUtil {
```
Comment (lioncash) [inline src/common/virtual_file_util.cpp]: There should be a newline after this.

File: src/common/virtual_file_util.cpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
+// Copyright 2019 yuzu Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include "virtual_file_util.h"
```
Comment (lioncash) [inline src/common/virtual_file_util.cpp]: Project includes should always be relative to the source root.

```suggestion
#include "common/virtual_file_util.h"
```

File: src/common/virtual_file_util.cpp [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+u8* const VirtualBinaryFile::Data() {
+    return m_file_storage.data();
+}
+
+std::size_t VirtualBinaryFile::Size() const {
+    return m_file_storage.size();
+}
+
+const void VirtualBinaryFile::Resize(std::size_t size) {
```
Comment (lioncash) [inline src/common/virtual_file_util.cpp]: This const isn't needed

File: src/common/virtual_file_util.cpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    m_file_storage_current_position = 0;
+}
+
+VirtualBinaryFile::~VirtualBinaryFile() {}
+
+void VirtualBinaryFile::ResetReadWritePosition() {
+    m_file_storage_current_position = 0;
+}
+
+u8* const VirtualBinaryFile::Data() {
```
Comment (lioncash) [inline src/common/virtual_file_util.cpp]: Ditto

File: src/common/virtual_file_util.h [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+#pragma once
+
+#include "common/common_types.h"
+
+#include <vector>
+
+namespace Common::VirtualFileUtil {
+
+class VirtualBinaryFile {
```
Comment (lioncash) [inline src/common/virtual_file_util.h]: Missing a documentation comment explaining its purpose.

File: src/common/virtual_file_util.h [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+    /// @return the current position of the virtual file on which the next read or write request
+    /// will start
+    std::size_t GetCurrentReadWritePosition() const;
+
+    /// set the current position of the virtual file on which the next read or write request
+    /// will start to the start of the virtual file (position = 0)
+    void ResetReadWritePosition();
+
+    /// @return pointer to the first byte of the virtual file
```
Comment (lioncash) [inline src/common/virtual_file_util.h]:
```suggestion
    /// @return pointer to the virtual file data
```

File: src/common/virtual_file_util.h [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    u8* const Data();
+
+    /// @return size in bytes of the virtual file
```
Comment (lioncash) [inline src/common/virtual_file_util.h]:
```suggestion
    /// @return virtual file size in bytes
```

File: src/common/zstd_compression.cpp [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+#include <algorithm>
+#include <zstd.h>
+
+#include "common/assert.h"
+#include "common/zstd_compression.h"
+
+namespace Common::Compression {
+
+std::vector<u8> CompressDataZSTD(const u8* source, std::size_t source_size, s32 compression_level) {
+
```
Comment (lioncash) [inline src/common/zstd_compression.cpp]: Unnecessary newline

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 12]
```diff
@@ -243,8 +211,17 @@ ShaderDiskCacheOpenGL::LoadPrecompiled() {
 std::optional<std::pair<std::unordered_map<u64, ShaderDiskCacheDecompiled>,
                         std::unordered_map<ShaderDiskCacheUsage, ShaderDiskCacheDump>>>
 ShaderDiskCacheOpenGL::LoadPrecompiledFile(FileUtil::IOFile& file) {
+
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp]: Unnecessary newline

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 13]
```diff
@@ -367,93 +330,112 @@ std::optional<ShaderDiskCacheDecompiled> ShaderDiskCacheOpenGL::LoadDecompiledEn
...
+            static_cast<u32>(PrecompiledEntryKind::Decompiled)) != 1 ||
+        m_precompiled_cache_virtual_file.WriteObject(unique_identifier) != 1 ||
+        m_precompiled_cache_virtual_file.WriteObject(static_cast<u32>(code.size())) != 1 ||
+        m_precompiled_cache_virtual_file.WriteArray(code.data(), code.size()) != code.size()) {
         return false;
     }
 
-    if (file.WriteObject(static_cast<u32>(entries.const_buffers.size())) != 1)
+    if (m_precompiled_cache_virtual_file.WriteObject(
+            static_cast<u32>(entries.const_buffers.size())) != 1)
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp]: This if statement body should be braced

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 14]
```diff
@@ -367,93 +330,112 @@ std::optional<ShaderDiskCacheDecompiled> ShaderDiskCacheOpenGL::LoadDecompiledEn
...
+            m_precompiled_cache_virtual_file.WriteObject(static_cast<u32>(cbuf.GetIndex())) != 1 ||
+            m_precompiled_cache_virtual_file.WriteObject(
+                static_cast<u8>(cbuf.IsIndirect() ? 1 : 0)) != 1) {
             return false;
         }
     }
 
-    if (file.WriteObject(static_cast<u32>(entries.samplers.size())) != 1)
+    if (m_precompiled_cache_virtual_file.WriteObject(static_cast<u32>(entries.samplers.size())) !=
+        1)
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp]: Ditto

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 15]
```diff
@@ -367,93 +330,112 @@ std::optional<ShaderDiskCacheDecompiled> ShaderDiskCacheOpenGL::LoadDecompiledEn
...
+                static_cast<u8>(sampler.IsArray() ? 1 : 0)) != 1 ||
+            m_precompiled_cache_virtual_file.WriteObject(
+                static_cast<u8>(sampler.IsShadow() ? 1 : 0)) != 1) {
             return false;
         }
     }
 
-    if (file.WriteObject(static_cast<u32>(entries.global_memory_entries.size())) != 1)
+    if (m_precompiled_cache_virtual_file.WriteObject(
+            static_cast<u32>(entries.global_memory_entries.size())) != 1)
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp]: Ditto

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 16]
```diff
@@ -367,93 +330,112 @@ std::optional<ShaderDiskCacheDecompiled> ShaderDiskCacheOpenGL::LoadDecompiledEn
...
+                1 ||
+            m_precompiled_cache_virtual_file.WriteObject(static_cast<u32>(gmem.GetCbufOffset())) !=
+                1) {
             return false;
         }
     }
 
     for (const bool clip_distance : entries.clip_distances) {
-        if (file.WriteObject(static_cast<u8>(clip_distance ? 1 : 0)) != 1)
+        if (m_precompiled_cache_virtual_file.WriteObject(static_cast<u8>(clip_distance ? 1 : 0)) !=
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp]: Ditto

File: src/video_core/renderer_opengl/gl_shader_disk_cache.h [hunk 17]
```diff
@@ -238,6 +242,10 @@ class ShaderDiskCacheOpenGL {
     Core::System& system;
     // Stored transferable shaders
     std::map<u64, std::unordered_set<ShaderDiskCacheUsage>> transferable;
+    // stores whole precompiled cache which will be read from or saved to the precompiled chache
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.h]:
```suggestion
    // Stores whole precompiled cache which will be read from or saved to the precompiled cache
```

**Comment by jroweboy** [comment]
Don't have time for a review ATM, but could you post a small benchmark difference between lz4 with your whole file compression and zstd with whole file compression? I am a little nervous about adding a new dependency to core (through common) if it ends up that lz4 on the whole file is roughly the same.

**Comment by ReinUsesLisp** [comment]
Does your benchmark include compilation times or just the time that it takes to save the shaders? From what you've posted in discord it takes 4.41 seconds to build and save the precompiled cache, where the save just takes <1 second.

**Comment by FreddyFunk** [comment]
@ReinUsesLisp this gets benchmarked as well

> Also the precompiled gets generated completly in memory with the help of a `virtual binary file`. This eleminates the bottleneck of disk latency and improves speed greatly.

I benchmarked the call of this function:
https://github.com/yuzu-emu/yuzu/blob/master/src/video_core/renderer_opengl/gl_rasterizer.cpp#L484



@jroweboy Here are some quick numbers. Remember the super fast NVME SSD I'm using. On a system with a slower disk LZ4 might be slower due to the bigger file size. Also my cache file is small, so please judge on precentage differences.

| Compression Algorithm | Compression Level | Size (KB)  | Precompiled shader cache file generation time (seconds)  |
| ------------- |:-------------:| -----:| -----:|
| uncompressed | --- | 79271 |  15.33 |
| **ZStandard** | 3 | **8809** | (currently in this PR) **4.41** |
| LZ4 | --- | 18461 | 4.37 |
| LZ4HC | 9 | 11885 |  6.19 |

File: src/common/virtual_file_util.cpp (hunk 4 above)
Comment (DarkLordZach) [inline src/common/virtual_file_util.cpp]: This entire namespace is quite redundant, yuzu already has a virtual file system implementation in core/file_sys. Please use that instead.

**Comment by ReinUsesLisp** [comment]
> @ReinUsesLisp all benchmarks target the whole generation of the precompiled shader cache file (including compression and saving to disk).

That means you are not doing a proper benchmark. If you want to improve the time it takes to **save** the cache, you have to benchmark the function that saves the precompiled cache. `LoadDiskCache` includes more than just saving the precompiled cache and some of that is inconsistent across executions.

**Comment by FernandoS27** [comment]
This PR appears to cause segmentation faults in linux.

**Comment by FreddyFunk** [comment]
I'm really busy this week, but I'll address the comments as soon as possible.

**Comment by FreddyFunk** [comment]
@DarkLordZach Thanks for the suggestion. I used a the VectorVfsFile class as a base now.

@FernandoS27 segfault should be fixed now

@ReinUsesLisp I think you missunderstood something. The benchmark are not **only** about the time it takes to "save" or "compress" the precompiled file. As the title of this PR says, it is about the time to **generate** the precompiled shader cache file. `LoadDiskCache` is the closest point I could find to measure time for the generation on a single point. Also please keep in mind, that these benchmarks are just a quick test and not an scientific evaluation.

File: src/core/file_sys/vfs_vector.h [hunk 18]
```diff
@@ -80,12 +80,62 @@ class VectorVfsFile : public VfsFile {
...
 private:
-    std::vector<u8> data;
     VirtualDir parent;
     std::string name;
 };
 
+// An implementation of VfsFile in append mode that is backed by a vector optionally supplied upon
+// construction. The advantage from this class over the VectorVfsFile is that no external tracking
+// of any offsets is required.
+class VectorVfsFileAppendMode : public VectorVfsFile {
```
Comment (DarkLordZach) [inline src/core/file_sys/vfs_vector.h]: Sorry, but personally I disagree with the design choice of adding `*Append` methods to an extension of the class. Not only does this pollute the interface, but it also is misleading if someone isn't informed on the class. Say perhaps someone wants a VFS class that works by appending and so uses `VectorVfsFileAppendMode` and then calls `Write` multiple times (with no offset)? Also this does nothing for overflow. Personally, given its only one use case I really don't see why you cant just keep track of offset at the use site.

File: src/common/virtual_file_util.h [hunk 19]
```diff
@@ -0,0 +1,111 @@
+// Copyright 2019 yuzu Emulator Project
```
Comment (DarkLordZach) [inline src/common/virtual_file_util.h]: Why is this (and associated cpp file) still in the PR? From what I've seen they aren't used anywhere anymore.

**Comment by FreddyFunk** [comment]
@bunnei ping for canary tagging

File: src/core/file_sys/vfs_vector.cpp [hunk 20]
```diff
@@ -8,7 +8,7 @@
 
 namespace FileSys {
 VectorVfsFile::VectorVfsFile(std::vector<u8> initial_data, std::string name, VirtualDir parent)
-    : data(std::move(initial_data)), parent(std::move(parent)), name(std::move(name)) {}
```
Comment (DarkLordZach) [inline src/core/file_sys/vfs_vector.cpp]: Please don't change random field names to match your style. 'm_' for member fields isn't part of our style guide and one of the points of the guide is that you should attempt to match style with adjacent code.

File: src/core/file_sys/vfs_vector.h [hunk 21]
```diff
@@ -66,7 +66,7 @@ class VectorVfsFile : public VfsFile {
 public:
     explicit VectorVfsFile(std::vector<u8> initial_data = {}, std::string name = "",
                            VirtualDir parent = nullptr);
-    ~VectorVfsFile() override;
```
Comment (DarkLordZach) [inline src/core/file_sys/vfs_vector.h]: why?

File: src/core/file_sys/vfs_vector.h [hunk 22]
```diff
@@ -80,12 +80,62 @@ class VectorVfsFile : public VfsFile {
 
     virtual void Assign(std::vector<u8> new_data);
 
+protected:
```
Comment (DarkLordZach) [inline src/core/file_sys/vfs_vector.h]: why? This class works under the assumption that data is private state.

File: src/core/file_sys/vfs_vector.h (hunk 18 above)
Comment (FreddyFunk) [reply src/core/file_sys/vfs_vector.h]: The original  `gl_shader_disk_cache.cpp` has a lot of code that looks like this. A lot of the logic is build around that the file (which is now being virtualized) is opend in append mode, which the current VFS does not support.  

``` cpp
if (file.WriteObject(static_cast<u32>(PrecompiledEntryKind::Dump)) != 1 ||
        file.WriteObject(usage) != 1 || file.WriteObject(static_cast<u32>(binary_format)) != 1 ||
        file.WriteObject(static_cast<u32>(binary_length)) != 1 ||
        file.WriteObject(static_cast<u32>(compressed_binary.size())) != 1 ||
        file.WriteArray(compressed_binary.data(), compressed_binary.size()) !=
            compressed_binary.size()) {
        LOG_ERROR(Render_OpenGL, "Failed to save binary program file in shader={:016x} - removing",
                  usage.unique_identifier);
        file.Close();
        InvalidatePrecompiled();
        return;
    }
```

For simplicity reasons I choosed to implemend a virtual file, that could be opened in append mode as well. I see your point with the potential missleading interface of the resulting `VectorVfsFileAppendMode` class. Therefore I followed your suggestion and used the plain `VectorVfsFile` class instead and tracked the offset within the `gl_shader_disk_cache`. Now the same code looks like this:

``` cpp
if (!SaveObjectToPrecompiled(static_cast<u32>(PrecompiledEntryKind::Dump)) ||
    !SaveObjectToPrecompiled(usage) ||
    !SaveObjectToPrecompiled(static_cast<u32>(binary_format)) ||
    !SaveObjectToPrecompiled(static_cast<u32>(binary_length)) ||
    !SaveArrayToPrecompiled(binary.data(), binary.size())) {
    LOG_ERROR(Render_OpenGL, "Failed to save binary program file in shader={:016x} - removing",
                usage.unique_identifier);
    InvalidatePrecompiled();
    return;
}
```

File: src/common/virtual_file_util.h (hunk 19 above)
Comment (FreddyFunk) [reply src/common/virtual_file_util.h]: git history is cleaned up now

**Comment by bunnei** [comment]
@FreddyFunk can you rebase? Then we can canary tag. Thank you!

**Comment by bunnei** [comment]
@FreddyFunk ping again to rebase when you get a chance, thank you.

**Comment by FreddyFunk** [comment]
@bunnei Rebased

File: src/video_core/renderer_opengl/gl_shader_cache.cpp [hunk 23]
```diff
@@ -524,4 +534,4 @@ Shader ShaderCacheOpenGL::GetStageProgram(Maxwell::ShaderProgram program) {
     return last_shaders[static_cast<u32>(program)] = shader;
 }
 
-} // namespace OpenGL
+} // namespace OpenGL
```
Comment (jroweboy) [inline src/video_core/renderer_opengl/gl_shader_cache.cpp]: super nit: re add the new line

File: src/video_core/renderer_opengl/gl_shader_disk_cache.cpp [hunk 24]
```diff
@@ -632,4 +623,4 @@ std::string ShaderDiskCacheOpenGL::GetTitleID() const {
     return fmt::format("{:016X}", system.CurrentProcess()->GetTitleID());
 }
 
-} // namespace OpenGL
+} // namespace OpenGL
```
Comment (jroweboy) [inline src/video_core/renderer_opengl/gl_shader_disk_cache.cpp]: super nit: ditto

**Comment by FreddyFunk** [comment]
With this PR the precompiled cache gets only written to disk after all shaders are compiled. By that I mean after the progress bars are finished and before the loading of the game happens.

**Pull request merged**